File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        last_lower = {}
        first_upper = {}
        
        for c in word:
            if c.islower():
                last_lower[c] = word.rindex(c)
            elif c.isupper() and c.lower() not in first_upper:
                first_upper[c.lower()] = word.index(c)
        
        return sum(
            1 for c in last_lower
            if c in first_upper and last_lower[c] < first_upper[c]
        ) 

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.rs
use std::collections::HashMap;

impl Solution {
    pub fn number_of_special_chars(word: String) -> i32 {
        let mut last_lower: HashMap<char, usize> = HashMap::new();
        let mut first_upper: HashMap<char, usize> = HashMap::new();

        for (i, c) in word.chars().enumerate() {
            if c.is_lowercase() {
                last_lower.insert(c, i);
            } else if c.is_uppercase() {
                let lc = c.to_lowercase().next().unwrap();
                first_upper.entry(lc).or_insert(i);
            }
        }

        last_lower.iter().filter(|(c, li)| {
            first_upper.get(c).map_or(false, |ui| *li < ui)
        }).count() as i32
    }
}